### PR TITLE
[Feature-fix] Make it so get_connected_nodes only returns not-deleted nodes

### LINKED ISSUE #3594: User Settings - Authorized Projects list in Mendeley & Zotero do not clear out deleted projects
### Steps
1. Go to [osf.io](http://www.osf.io)
2. Create a new project or fork an existing project
3. Configure and enable Mendeley and Zotero, as well as a storage add-on, in your project
4. Go to your User Settings page
5. Go to the Configure Add-ons page
6. Confirm that the new project appears as an Authorized Project under Mendeley, Zotero, and the storage add-on you chose
7. Go to the new project's Settings page
8. Delete the project
9. Go back to the Configure Add-ons page

### Expected behavior
I expected that the Configure Add-ons page would not display the deleted project anywhere.

### Actual behavior
Although the storage add-ons clear the deleted project from their Authorized Projects lists, Mendeley and Zotero still display it.

In this example, both "Fork of Benediction in the morning" (only enabled Mendeley) and "Tester" have been deleted, but they are still shown under in Mendeley and Zotero. If I click the link to them, I arrive at the Resource Deleted page.

screen shot 2015-07-14 at 1 53 36 pm

### PR DESCRIPTION
**Pull request by hmoco**
## Purpose
Fixes #3594. Prevents Zotero and Mendeley from listing deleted nodes under their list of connected nodes in user settings. 

## Changes
Make it so get_connected_nodes only returns not-deleted nodes

## Side-Effects
Don't think there will be any, this seems to only be used by the update_accounts method that gets inherited and used by the citations, where it fixes the problem

### CODE REVIEW

File: website/addons/base/__init__.py [hunk 1]
```diff
@@ -602,6 +602,8 @@ def get_attached_nodes(self, external_account):
         for node in self.get_nodes_with_oauth_grants(external_account):
             if node is None:
                 continue
+            if node.is_deleted:
```
Comment (sloria) [inline website/addons/base/__init__.py]: It might make sense to filter out deleted nodes in `get_nodes_with_oauth_grants` instead of here. Are there any places we call that method where we would want to include deleted nodes?

**Comment by sloria** [comment]
Pass finished

File: website/addons/base/__init__.py (hunk 1 above)
Comment (hmoco) [inline website/addons/base/__init__.py]: Nope, it seems to only be used by ```get_attached_nodes```

Comment (hmoco) [inline website/addons/base/__init__.py]: Filtering out can be done while ```get_nodes_with_oauth_grants``` is iterating through ```oauth_grants```

Comment (sloria) [inline website/addons/base/__init__.py]: Yes, I think that would make more sense; that way, we don't need to repeat this check if we call `get_nodes_with_oauth_grants` elsewhere.

File: website/addons/base/__init__.py [hunk 2]
```diff
@@ -595,7 +595,7 @@ def get_nodes_with_oauth_grants(self, external_account):
         return (
             Node.load(node_id)
             for node_id, grants in self.oauth_grants.iteritems()
-            if external_account._id in grants.keys()
+            if (external_account._id in grants.keys() and not Node.load(node_id).is_deleted)
```
Comment (hmoco) [inline website/addons/base/__init__.py]: Made this change. @sloria is it safe to load the node_id just to check if it is deleted?

**Comment by hmoco** [comment]
Also, not sure why this function, on the same file, wasn't taking care of that:
```
def on_delete(self):
        """When the user deactivates the addon, clear auth for connected nodes.
        """
        super(AddonOAuthUserSettingsBase, self).on_delete()
        nodes = [Node.load(node_id) for node_id in self.oauth_grants.keys()]
        for node in nodes:
            node_addon = node.get_addon(self.oauth_provider.short_name)
            if node_addon and node_addon.user_settings == self:
                node_addon.clear_auth()
```

File: website/addons/base/__init__.py (hunk 2 above)
Comment (sloria) [inline website/addons/base/__init__.py]: We already load the node on ~L596, so it's duplicative to load it here, too. So you might want to refactor this into a `for` loop, instead.

File: website/addons/base/__init__.py [hunk 3]
```diff
@@ -592,11 +592,13 @@ def verify_oauth_access(self, node, external_account, metadata=None):
 
     def get_nodes_with_oauth_grants(self, external_account):
         # Generator of nodes which have grants for this external account
-        return (
-            Node.load(node_id)
-            for node_id, grants in self.oauth_grants.iteritems()
-            if external_account._id in grants.keys()
-        )
+        nodes = []
+        for node_id, grants in self.oauth_grants.iteritems():
```
Comment (sloria) [inline website/addons/base/__init__.py]: Why was this changed to a loop. Why not just add the check for `not node.is_deleted` to the generator expression?

Comment (sloria) [inline website/addons/base/__init__.py]: Disregard. This should be a for loop to avoid having to load each Node twice.

Comment (sloria) [inline website/addons/base/__init__.py]: However, you can still make this method a generator by `yield`-ing each result.

**Comment by sloria** [comment]
Merged into develop.
